feat: Introduce Email and SMS delivery ingredients

KShivendu opened the pull request:
## Summary of change

Emailing ingredient for python SDK

## Test Plan


## Documentation changes

TODO

## Checklist for important updates

-   [x] Changelog has been updated
-   [ ] `coreDriverInterfaceSupported.json` file has been updated (if needed)
    -   Along with the associated array in `supertokens_python/constants.py`
-   [ ] `frontendDriverInterfaceSupported.json` file has been updated (if needed)
-   [ ] Changes to the version if needed
    -   In `setup.py`
    -   In `supertokens_python/constants.py`
-   [ ] Had installed and ran the pre-commit hook
-   [ ] Issue this PR against the latest non released version branch.
    -   To know which one it is, run find the latest released tag (`git tag`) in the format `vX.Y.Z`, and then find the latest branch (`git branch --all`) whose `X.Y` is greater than the latest released tag.
    -   If no such branch exists, then create one from the latest released branch.
-   [ ] If have added a new web framework, update the `supertokens_python/utils.py` file to include that in the `FRAMEWORKS` variable
-   [ ] If added a new recipe that has a User type with extra info, then be sure to change the User type in supertokens_python/types.py
 
## TODO:
- [x] logging for emails and errors from them
- [x] humanise time in email and sms
- [x] expose functions to send_sms and send_email
- [x] https://github.com/supertokens/supertokens-node/pull/321/files
- [x] convert `log_debug_inf(".... %s", exception)` to `log_debug_inf(".... %s", str(exception))`
- [x] Fix `email_delivery_ingredient = ingredients.email_delivery if ingredients else None` (Now ingredient cannot be None). 
- [x] Rename `password_reset_implementation.py` to `password_reset.py`
- [x] In `SMTPService` from `<recipe>/emaildelivery/services/smtp/__init__.py`, don't assign `self.config`
- [x] https://github.com/supertokens/supertokens-python/pull/100/files#r888664809
- [x] Finalize if we should share SMTP connections https://github.com/supertokens/supertokens-python/pull/100#discussion_r855885129
- [x] Finalize whether we want `user_context` inside `TypeXEmailDelivery` or not? (and do the same thing in all recipes)
- [x] Remove `async` and redundant `user_context = {}` from `asyncio/syncio` of recipes.

rishabhpoddar commented on supertokens_python/ingredients/emaildelivery/service/smtp.py:
```diff
@@ -0,0 +1,112 @@
+
+SmtpAuth = TypedDict('auth', {'user': str, 'password': str})
+
+SMTPServiceConfig = TypedDict('SMTPServiceConfig', {
+    'host': str,
+    'from': TypedDict('from', {'name': str, 'email': str}),
+    'port': int,
+    'secure': Union[bool, None],
+    'auth': Union[SmtpAuth, None],
+})
```
```suggestion
class SMTPServiceConfig
   def __init__(host: str, emailFrom: SMTPServiceConfigFrom, port: int, secure:  Union[bool, None] = None, auth: Union[SmtpAuth, None] = None)
      self.host = host
      self...

class SMTPServiceConfigFrom:
   def __init__(name: str, email: str):
      this.name = name
      this.email = email
```

KShivendu replied:
Done.

rishabhpoddar commented on supertokens_python/ingredients/emaildelivery/service/smtp.py:
```diff
@@ -0,0 +1,112 @@
+from typing import Any, Callable, Dict, Generic, TypedDict, TypeVar, Union
+
+_T = TypeVar('_T')
+
```
Fix all the types in this file please.

KShivendu replied:
Done.

rishabhpoddar commented on supertokens_python/recipe/emailverification/interfaces.py:
```diff
@@ -224,6 +227,17 @@ def __init__(self):
+class TypeEmailVerificationEmailDeliveryInput:
+    def __init__(
+        self, user: User,
+        email_verify_link: str, user_context: Dict[str, Any]
```
`user_context` should be a separate param and not as a field in this type

KShivendu replied:
todo

KShivendu replied:
Fixed.

rishabhpoddar commented on supertokens_python/recipe/emailpassword/emaildelivery/service/backward/__init__.py:
```diff
@@ -0,0 +1,55 @@
+            )
+
+            if user is None:
+                raise Exception("Shouldn't have come here")
```
```suggestion
                raise Exception("Should never come here")
```

KShivendu replied:
Fixed.

rishabhpoddar commented on supertokens_python/recipe/emailpassword/recipe.py:
```diff
@@ -49,22 +51,31 @@
 class EmailPasswordRecipe(RecipeModule):
     recipe_id = 'emailpassword'
     __instance = None
+    email_delivery: Union[EmailDeliveryIngredient[TypeEmailPasswordEmailDeliveryInput], None]
```
This should never be none

KShivendu replied:
Fixed.

rishabhpoddar commented on supertokens_python/recipe/emailpassword/recipe.py:
```diff
@@ -132,10 +143,12 @@ def get_all_cors_headers(self) -> List[str]:
              email_verification_feature: Union[InputEmailVerificationConfig, None] = None,
-             override: Union[InputOverrideConfig, None] = None):
+             override: Union[InputOverrideConfig, None] = None,
+             email_delivery_ingredient: Union[EmailDeliveryIngredient, None] = None,
+             ):
```
Also call it `email_derlivery`

KShivendu replied:
Fixed.

rishabhpoddar commented on supertokens_python/recipe/emailpassword/recipe.py:
```diff
@@ -49,22 +51,31 @@
                  sign_up_feature: Union[InputSignUpFeature, None] = None,
                  reset_password_using_token_feature: Union[InputResetPasswordUsingTokenFeature, None] = None,
                  email_verification_feature: Union[InputEmailVerificationConfig, None] = None,
                  override: Union[InputOverrideConfig, None] = None,
```
This should also take `email_delivery: Union[EmailDeliveryConfig< TypeEmailPasswordEmailDeliveryInput>, None] = None` too.

KShivendu replied:
Fixed

rishabhpoddar commented on supertokens_python/recipe/emailpassword/utils.py:
```diff
@@ -288,12 +292,15 @@ def __init__(self,
                  email_verification_feature: ParentRecipeEmailVerificationConfig,
-                 override: OverrideConfig):
+                 override: OverrideConfig,
+                 email_delivery: Union[TypeInput[TypeEmailPasswordEmailDeliveryInput], None] = None
```
```suggestion
                 email_delivery: Callable[All the stuff that the backward compatibility service needs], TypeInput[TypeEmailPasswordEmailDeliveryInput]]
```

KShivendu replied:
Done.

rishabhpoddar commented on supertokens_python/recipe/emailpassword/utils.py:
```diff
@@ -302,14 +309,29 @@ def validate_and_normalise_user_input(recipe: EmailPasswordRecipe, app_info: App
     if sign_up_feature is None:
         sign_up_feature = InputSignUpFeature()
+
+    email_service = email_delivery.service if email_delivery_config is not None else None
+    if email_service is None:
+        # assert config.create_and_send_custom_email is not None, "Shouldn't have happened"
+        email_service = BackwardCompatibilityService(
+            app_info,
+            # reset_password_using_token_feature,
+            # email_verification_feature,
+            # email_delivery_config
+        )
+
+        email_delivery = TypeInput[TypeEmailPasswordEmailDeliveryInput](email_service)
+
```
All of this will go inside a function and you return that function

KShivendu replied:
Yeah. Fixed

rishabhpoddar commented on supertokens_python/recipe/emailpassword/types.py:
```diff
@@ -55,3 +58,23 @@ def __init__(self, id: str, validate: Callable[[  # pylint: disable=redefined-bu
+
+class TypeEmailPasswordPasswordResetEmailDeliveryInput:
+    def __init__(self, user: User, password_reset_link: str, user_context: Dict[str, Any]) -> None:
+        self.type: str = "PASSWORD_RESET"
```
Don't need this type var

KShivendu replied:
Removed `self.type`

rishabhpoddar commented on supertokens_python/recipe/emailpassword/types.py:
```diff
@@ -55,3 +58,23 @@ def __init__(self, id: str, validate: Callable[[  # pylint: disable=redefined-bu
+
+
+class TypeEmailPasswordPasswordResetEmailDeliveryInput:
+    def __init__(self, user: User, password_reset_link: str, user_context: Dict[str, Any]) -> None:
```
User type will have to change

KShivendu replied:
Created `TypeEmailPasswordPasswordResetEmailDeliveryInputUser`

rishabhpoddar commented on supertokens_python/recipe/emailverification/api/implementation.py:
```diff
@@ -69,9 +70,12 @@ async def generate_email_verify_token_post(self, api_options: APIOptions, user_c
+            email_verify_link=email_verify_link,
+            user_context=user_context
+        )
+        assert api_options.email_delivery is not None
```
You don't need this assert anymore as it will always be true

KShivendu replied:
Done.

rishabhpoddar commented on supertokens_python/recipe/emailverification/interfaces.py:
```diff
@@ -125,12 +127,13 @@ async def unverify_email(self, user_id: str, email: str, user_context: Dict[str,
 class APIOptions:
     def __init__(self, request: BaseRequest, response: BaseResponse, recipe_id: str,
-                 config: EmailVerificationConfig, recipe_implementation: RecipeInterface):
+                 config: EmailVerificationConfig, recipe_implementation: RecipeInterface, email_delivery: Union[EmailDeliveryIngredient[TypeEmailVerificationEmailDeliveryInput], None]):
```
Should not be nullable

KShivendu replied:
Yeah Fixed.

rishabhpoddar commented on supertokens_python/recipe/emailverification/utils.py:
```diff
@@ -57,42 +64,49 @@ class ParentRecipeEmailVerificationConfig:
-                     User, str, Dict[str, Any]], Awaitable[None]], None] = None
+                 get_email_verification_url: Union[Callable[[User, Dict[str, Any]], Awaitable[str]], None] = None,
+                 create_and_send_custom_email: Union[Callable[[User, str, Dict[str, Any]], Awaitable[None]], None] = None,
+                 email_delivery: Union[TypeInput[TypeEmailVerificationEmailDeliveryInput], None] = None
                  ):
```
This is not needed.

KShivendu replied:
Fixed.

rishabhpoddar commented on supertokens_python/recipe/emailverification/utils.py:
```diff
@@ -57,42 +64,49 @@ class ParentRecipeEmailVerificationConfig:
-                     User, Dict[str, Any]], Awaitable[str]], None] = None,
-                 create_and_send_custom_email: Union[Callable[[
-                     User, str, Dict[str, Any]], Awaitable[None]], None] = None
+                 get_email_verification_url: Union[Callable[[User, Dict[str, Any]], Awaitable[str]], None] = None,
+                 create_and_send_custom_email: Union[Callable[[User, str, Dict[str, Any]], Awaitable[None]], None] = None,
```
Mark these inputs as deprecated.

rishabhpoddar commented on supertokens_python/recipe/emailverification/emaildelivery/services/backward/__init__.py:
```diff
@@ -0,0 +1,37 @@
+
+    def __init__(self,
+                 app_info: AppInfo,
+                 create_and_send_custom_email: Callable[[User, str, Dict[str, Any]], Awaitable[None]]
+                 ) -> None:
```
```suggestion
                 create_and_send_custom_email: Union[Callable[[User, str, Dict[str, Any]], Awaitable[None]]
                 ) -> None, None]:
```

KShivendu replied:
Done.

rishabhpoddar commented on supertokens_python/recipe/emailverification/utils.py:
```diff
@@ -57,42 +64,49 @@ class ParentRecipeEmailVerificationConfig:
+
+    email_service = config.email_delivery.service if config.email_delivery is not None else None
+    if email_service:
+        assert config.create_and_send_custom_email is not None, "Shouldn't have happened"
```
This assert should not be needed

KShivendu replied:
Yeah. Removed.

rishabhpoddar commented on supertokens_python/recipe/emailverification/utils.py:
```diff
@@ -57,42 +64,49 @@ class ParentRecipeEmailVerificationConfig:
-    create_and_send_custom_email = config.create_and_send_custom_email if config.create_and_send_custom_email is not None \
-        else default_create_and_send_custom_email(app_info)
+
+    email_service = config.email_delivery.service if config.email_delivery is not None else None
+    if email_service:
+        assert config.create_and_send_custom_email is not None, "Shouldn't have happened"
+        email_service = BackwardCompatibilityService(app_info, config.create_and_send_custom_email)
+
```
This whole thing goes in a function, and you return that function from this normalisation function

KShivendu replied:
Yeah. Fixed.

rishabhpoddar commented on supertokens_python/recipe/emailverification/utils.py:
```diff
@@ -57,42 +64,49 @@ class ParentRecipeEmailVerificationConfig:
-                 create_and_send_custom_email: Callable[[User, str, Dict[str, Any]], Awaitable[None]],
-                 get_email_for_user_id: Callable[[str, Dict[str, Any]], Awaitable[str]]
+                 get_email_for_user_id: Callable[[str, Dict[str, Any]], Awaitable[str]],
+                 email_delivery: Union[TypeInput[TypeEmailVerificationEmailDeliveryInput], None],
+                 email_service: Union[EmailDeliveryInterface[TypeEmailVerificationEmailDeliveryInput], None]
```
this should just be one of them and it it's type should be `Callable[[All things needed by backward compatible service], TypeInput[TypeEmailVerificationEmailDeliveryInput]]`

KShivendu replied:
Fixed.

rishabhpoddar commented on supertokens_python/recipe/emailverification/recipe.py:
```diff
@@ -57,6 +61,7 @@ def __init__(self, recipe_id: str, app_info: AppInfo,
         api_implementation = APIImplementation()
         self.api_implementation = api_implementation if self.config.override.apis is None else \
             self.config.override.apis(api_implementation)
+        self.email_delivery = EmailDeliveryIngredient(self.config.email_delivery) if (email_delivery_ingredient is None and self.config.email_delivery is not None) else email_delivery_ingredient
```
This needs to change too

KShivendu replied:
Fixed.

KShivendu commented on supertokens_python/ingredients/emaildelivery/types.py:
```diff
@@ -0,0 +1,35 @@
+
+class EmailDeliveryInterface(ABC, Generic[_T]):
+    @abstractmethod
+    def send_email(self, email_input: Union[_T, None]) -> Any:
```
Done

KShivendu commented on supertokens_python/ingredients/emaildelivery/types.py:
```diff
@@ -0,0 +1,35 @@
+        return original_impl
+
+    def __init__(self, service: EmailDeliveryInterface[_T]) -> None:
+        self.service = service
```
Done

KShivendu commented on supertokens_python/ingredients/emaildelivery/service/smtp.py:
```diff
@@ -0,0 +1,115 @@
+# Copyright (c) 2021, VRAI Labs and/or its affiliates. All rights reserved.
+#
+# This software is licensed under the Apache License, Version 2.0 (the
+# "License") as published by the Apache Software Foundation.
+#
+# You may not use this file except in compliance with the License. You may
+# obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+# License for the specific language governing permissions and limitations
+# under the License.
+
+
+from ..types import EmailDeliveryInterface
```
@rishabhpoddar please ignore this file while reviewing this time.

KShivendu replied:
Done.

KShivendu commented on supertokens_python/recipe/emailpassword/recipe.py:
```diff
@@ -49,26 +55,37 @@
+        else:
+            self.email_verification_recipe = EmailVerificationRecipe(recipe_id, app_info,
+                                                                     self.config.email_verification_feature,
+                                                                     email_delivery_ingredient=self.email_delivery_ingredient)
+
```
"TypeEmailPasswordPasswordResetEmailDeliveryInput" is incompatible with "TypeEmailVerificationEmailDeliveryInput"

bhumilsarvaiya commented on supertokens_python/recipe/emailverification/interfaces.py:
```diff
@@ -224,6 +228,21 @@ def __init__(self):
+        email_verify_link: str, user_context: Dict[str, Any]
+    ) -> None:
+        self.user = user
+        self.email_verify_link = email_verify_link
```
initialize `user_context`

KShivendu replied:
Fixed.

bhumilsarvaiya commented on supertokens_python/recipe/emailverification/emaildelivery/services/backwardCompatibility/__init__.py:
```diff
@@ -0,0 +1,34 @@
+
```
add license header

KShivendu replied:
Added.

bhumilsarvaiya commented on supertokens_python/recipe/emailverification/recipe.py:
```diff
@@ -45,9 +47,11 @@
     def __init__(self, recipe_id: str, app_info: AppInfo,
-                 config: ParentRecipeEmailVerificationConfig):
+                 config: ParentRecipeEmailVerificationConfig,
+                 email_delivery_ingredient: Union[EmailDeliveryIngredient[TypeEmailVerificationEmailDeliveryInput], None] = None):
```
just make ingredient class of which email_delivery will be a pert of

bhumilsarvaiya commented on supertokens_python/recipe/emailpassword/emaildelivery/service/backwardCompatibility/__init__.py:
```diff
@@ -0,0 +1,104 @@
+                reset_password_using_token_feature.create_and_send_custom_email = default_create_and_send_custom_email(self.app_info)
+            self.reset_password_using_token_feature = reset_password_using_token_feature
+
+        if email_verification_feature:
```
```python
if email_verification_feature.create_and_send_custom_email is not None:
    tmp = email_verification_feature.create_and_send_custom_email
    async def create_and_send_custom_email_wrapper(user: User, link: str, user_context: Dict[str, Any]):
        user_info = await self.recipeInterfaceImpl.recipe_implementation.get_user_by_id(user.user_id, user_context)
        if user_info is None:
            raise Exception("Unknown User ID provided")

        # assert email_verification_feature.create_and_send_custom_email is not None
        return await tmp(user, link, user_context)
```

bhumilsarvaiya commented on supertokens_python/recipe/emailpassword/emaildelivery/service/backwardCompatibility/__init__.py:
```diff
@@ -0,0 +1,104 @@
+                    user_info = await self.recipeInterfaceImpl.recipe_implementation.get_user_by_id(user.user_id, user_context)
+                    if user_info is None:
+                        raise Exception("Unknown User ID provided")
+
```
you can do this instead of assert and cast
```python
create_and_send_custom_email: Union[Callable[[
                EmailVerificationUser, str, Dict[str, Any]], Awaitable[None]], None] = None
if email_verification_feature:
    if email_verification_feature.create_and_send_custom_email is not None:
        tmp = email_verification_feature.create_and_send_custom_email
        async def create_and_send_custom_email_wrapper(user: EmailVerificationUser, link: str, user_context: Dict[str, Any]):
            user_info = await self.recipeInterfaceImpl.recipe_implementation.get_user_by_id(user.user_id, user_context)
            if user_info is None:
                raise Exception("Unknown User ID provided")

            return await tmp(user_info, link, user_context)

        create_and_send_custom_email = create_and_send_custom_email_wrapper

self.emailVerificationBackwardCompatiblityService = EmailVerificationBackwardCompatibilityService(
    app_info, create_and_send_custom_email=create_and_send_custom_email
)
```

bhumilsarvaiya commented on supertokens_python/recipe/emailpassword/emaildelivery/service/backwardCompatibility/__init__.py:
```diff
@@ -0,0 +1,104 @@
+
+
+class CreateAndSendCustomEmailInput:
+    user: str
```
```python
class ResetPasswordUsingTokenFeature:
    def __init__(self, create_and_send_custom_email: Callable[[User, str, Dict[str, Any]], Awaitable[None]]) -> None:
        self.create_and_send_custom_email = create_and_send_custom_email
```

bhumilsarvaiya commented on supertokens_python/ingredients/emaildelivery/email_delivery.py:
```diff
@@ -0,0 +1,35 @@
+# under the License.
+
+from typing import Any, Dict, Generic, TypeVar
+
```
move the content of this file to __init__.py and then remove this file

KShivendu commented:
Created https://github.com/supertokens/supertokens-python/pull/102/files for next iteration

rishabhpoddar commented:
@KShivendu @bhumilsarvaiya what is the progess on this?

What about the other recipes? When will this be done by?

KShivendu commented:
Some tests are failing. Please don't review. I'll look into the issue.

rishabhpoddar commented on supertokens_python/recipe/emailpassword/emaildelivery/service/smtp/service_implementation/email_verification_implementation.py:
```diff
@@ -0,0 +1,33 @@
+
+
+class ServiceImplementation(ServiceInterface[TypeEmailVerificationEmailDeliveryInput]):
+    def __init__(self, email_verification_service_implementation: ServiceInterface[TypeEmailPasswordEmailDeliveryInput]) -> None:
```
This variable should be called `email_password_service_implementation`? (See the type of it)

rishabhpoddar commented on supertokens_python/recipe/emailpassword/emaildelivery/service/smtp/service_implementation/__init__.py:
```diff
@@ -0,0 +1,55 @@
+        return get_password_reset_email_content(email_input)
+
+
+def getServiceImplementation(transporter: Transporter) -> ServiceInterface[TypeEmailPasswordEmailDeliveryInput]:
+    si = ServiceImplementation(transporter)
+    return si
```
This seems to be unused. Remove it?

rishabhpoddar replied:
Nope. I saw it, and thought I have not seen this anywhere else, so it must be unused. But there probably is some plugin out there to detect unused code

rishabhpoddar commented on supertokens_python/recipe/emailpassword/emaildelivery/service/smtp/password_reset_implementation.py:
```diff
@@ -0,0 +1,382 @@
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+# License for the specific language governing permissions and limitations
+# under the License.
+
```
Are the contents of this file in sync with node SDK?

rishabhpoddar commented on supertokens_python/recipe/emailpassword/emaildelivery/service/smtp/service_implementation/__init__.py:
```diff
@@ -0,0 +1,55 @@
+        email_verification_service_implementation.get_content = derived_ev_service_implementation.get_content
+
+    async def send_raw_email(self, get_content_result: GetContentResult, config_from: SMTPServiceConfigFrom, user_context: Dict[str, Any]) -> None:
+        self.transporter.send_email(config_from, get_content_result, user_context)
```
Are you sure that we don't have to await this? How will this behave in an event loop driven runtime env?

KShivendu replied:
You're right. Using `await` now.

rishabhpoddar commented on tests/test_email.py:
```diff
@@ -0,0 +1,273 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
```
Please write down all the existing implemented test cases in a checklist in the PR comment (in the "Test Plan" section)

KShivendu commented on supertokens_python/ingredients/emaildelivery/service/smtp.py:
```diff
@@ -0,0 +1,94 @@
+    def __init__(self, smtpSettings: SMTPServiceConfig) -> None:
+        self.smtpSettings = smtpSettings
+
+    def send_email(self, config_from: SMTPServiceConfigFrom, get_content_result: GetContentResult,
+                   _: Dict[str, Any]) -> None:
+        try:
+            smtp = smtplib.SMTP(self.smtpSettings.host, self.smtpSettings.port)
+            if self.smtpSettings.secure:
+                smtp.starttls()
+            if self.smtpSettings.auth:
+                smtp.login(self.smtpSettings.auth.user, self.smtpSettings.auth.password)
+
+            smtp.sendmail(config_from.email, get_content_result.to_email, get_content_result.body)
+            smtp.quit()
+        except Exception:
+            pass
```
- `semmail` throws [these errors](https://github.com/python/cpython/blob/main/Lib/smtplib.py#L838). How should I decide which one to consider?
- Valid point. Although it's always getting called by async function, we should make this function async as well.
- Not so sure about it. What if multiple emails are being sent at once using the same transporter object? @sattvikc what do you think?

sattvikc replied:
@KShivendu @rishabhpoddar Regarding the third point, I would not prefer sharing SMTP connections for sending multiple emails, unless we have proper queuing mechanisms in place for sending emails. Doing everything within send_email seems more reliable. smtp.close() also needs to be added I think, please check once.

rishabhpoddar replied:
@sattvikc any documentation from where you found this? How do you know that sharing one instance is bad?

sattvikc replied:
On the creation of SMTP class, the connection gets established. And by the time we need to send the email, there is a possibility that the connection might have timed out. So additional code will be required to check all those corner conditions.

sattvikc commented on supertokens_python/ingredients/emaildelivery/types.py:
```diff
@@ -0,0 +1,35 @@
+        pass
+
+
+class EmailDeliveryConfig(ABC, Generic[_T]):
+    def __init__(
+        self,
+        service: EmailDeliveryInterface[_T],
+        override: Union[Callable[[EmailDeliveryInterface[_T]], EmailDeliveryInterface[_T]], None] = None,
+    ) -> None:
```
make service optional, add another type with service mandatory

sattvikc commented on supertokens_python/ingredients/emaildelivery/service/smtp.py:
```diff
@@ -0,0 +1,97 @@
+
+class EmailDeliverySMTPConfig(Generic[_T]):
+    def __init__(self,
+                 smtpSettings: SMTPServiceConfig,
```
follow snake casing - smtp_settings

KShivendu replied:
Done. Thanks!

sattvikc commented on supertokens_python/recipe/emailpassword/emaildelivery/service/smtp/password_reset_implementation.py:
```diff
@@ -0,0 +1,657 @@
+    return content_result
+
+
+def get_password_reset_email_html(appName: str, email: str, resetLink: str):
+    template = Template("""
```
add html file in the same dir, reference it thru os.path.dirname(__file__)

KShivendu replied:
Done. Thanks!

sattvikc commented on supertokens_python/recipe/emailpassword/utils.py:
```diff
@@ -185,11 +172,13 @@ def __init__(self,
         self.form_fields_for_generate_token_form = form_fields_for_generate_token_form
         self.get_reset_password_url = get_reset_password_url
-        self.create_and_send_custom_email = create_and_send_custom_email
+        self.create_and_send_custom_email = None
```
you should still assign the argument instead of None. the option should work until we remove the parameter.

KShivendu replied:
done

sattvikc commented on supertokens_python/recipe/emailverification/emaildelivery/service/backwardCompatibility/__init__.py:
```diff
@@ -0,0 +1,52 @@
+# Copyright (c) 2021, VRAI Labs and/or its affiliates. All rights reserved.
```
change the directory from backwardCompatibility to backward_compatibility

KShivendu replied:
Done. Thanks!

sattvikc commented on supertokens_python/recipe/emailverification/emaildelivery/service/smtp/email_verify.py:
```diff
@@ -0,0 +1,657 @@
+    return content_result
+
+
+def get_email_verify_email_html(appName: str, email: str, verificationLink: str):
```
create html file

KShivendu replied:
Done. Thanks!

sattvikc commented on supertokens_python/recipe/emailverification/utils.py:
```diff
@@ -57,42 +52,56 @@ class ParentRecipeEmailVerificationConfig:
-        self.create_and_send_custom_email = create_and_send_custom_email
         self.get_email_for_user_id = get_email_for_user_id
+        self.email_delivery = email_delivery
+        self.create_and_send_custom_email = None
```
assgin the argument instead of None

KShivendu replied:
done.

sattvikc commented on supertokens_python/recipe/emailverification/utils.py:
```diff
@@ -57,42 +52,56 @@ class ParentRecipeEmailVerificationConfig:
-        self.create_and_send_custom_email = create_and_send_custom_email
         self.override = override
+        self.get_email_delivery_config = get_email_delivery_config
+        self.create_and_send_custom_email = None
```
assign the argument

KShivendu replied:
done.

sattvikc reviewed (changes requested):
Refer my comments

KShivendu commented on supertokens_python/recipe/emailpassword/emaildelivery/service/backwardCompatibility/__init__.py:
```diff
@@ -0,0 +1,120 @@
+# Copyright (c) 2021, VRAI Labs and/or its affiliates. All rights reserved.
```
Done. Thanks!

KShivendu commented:
Completed dev testing for the SMTP feature. It is working as expected.

KShivendu commented:
Hey @rishabhpoddar 
I observed that the HTML email content shows `IF:MC_PREVIEW_TEXT` (Commented HTML) which can feel spammy to the users. I think we should remove remove it.

KShivendu commented on tests/test_email.py:
```diff
@@ -0,0 +1,273 @@
+# Copyright (c) 2021, VRAI Labs and/or its affiliates. All rights reserved.
```
done

rishabhpoddar replied:
Not well enough!

sattvikc commented on supertokens_python/recipe/emailpassword/utils.py:
```diff
@@ -185,11 +172,13 @@ def __init__(self,
         self.get_reset_password_url = get_reset_password_url
         self.create_and_send_custom_email = create_and_send_custom_email
+        if create_and_send_custom_email:
+            warn("create_and_send_custom_email is depricated. Please use email delivery config instead")
```
deprecated (typo)

KShivendu replied:
Thanks!

sattvikc commented on supertokens_python/recipe/emailverification/emaildelivery/service/smtp/__init__.py:
```diff
@@ -0,0 +1,43 @@
+            self.config.smtp_settings.email_from.name,
+            self.config.smtp_settings.email_from.email
+        )
+        await self.serviceImpl.send_raw_email(content, send_raw_email_from, user_context)
```
`send_raw_email_from` can be removed

KShivendu replied:
Done.

sattvikc commented on supertokens_python/recipe/emailverification/emaildelivery/service/smtp/email_verify.py:
```diff
@@ -0,0 +1,38 @@
+def get_email_verify_email_html(appName: str, email: str, verificationLink: str):
+    current_dir = path.dirname(__file__)
+    template_path = path.join(current_dir, "email_verify_email.html")
+    template = open(template_path, "r").read()
```
File not closed. you could enclose it within `with` block

KShivendu replied:
Done. Thanks!

sattvikc commented on supertokens_python/recipe/emailverification/recipe.py:
```diff
@@ -45,9 +49,12 @@
     def __init__(self, recipe_id: str, app_info: AppInfo,
-                 config: ParentRecipeEmailVerificationConfig):
+                 config: ParentRecipeEmailVerificationConfig,
+                 ingredients: Union[EmailVerificationIngredients, None] = None
```
Do you want to call it `email_ingredient` ? We will have sms ingredient as well in few recipes

KShivendu replied:
`EmailVerificationIngredients` can contain `EmailDeliveryIngredient` as well `SMSDeliveryIngredient`. This is the same as the node PR.

sattvikc commented on supertokens_python/recipe/emailpassword/emaildelivery/service/smtp/password_reset_implementation.py:
```diff
@@ -0,0 +1,38 @@
+def get_password_reset_email_html(appName: str, email: str, resetLink: str):
+    current_dir = path.dirname(__file__)
+    template_path = path.join(current_dir, "password_reset_email.html")
+    template = open(template_path, "r").read()
```
File not closed, use `with` block.

KShivendu replied:
Done.

KShivendu commented on supertokens_python/ingredients/emaildelivery/service/smtp.py:
```diff
@@ -0,0 +1,103 @@
+    async def send_email(self, config_from: SMTPServiceConfigFrom, get_content_result: GetContentResult,
+                         _: Dict[str, Any]) -> None:
+        smtp = smtplib.SMTP(self.smtp_settings.host, self.smtp_settings.port)
+        try:
+            if self.smtp_settings.secure:
+                smtp.starttls()
+            if self.smtp_settings.auth:
+                smtp.login(self.smtp_settings.auth.user, self.smtp_settings.auth.password)
+
+            email_content = MIMEText(get_content_result.body, "html")
+            email_content["From"] = config_from.email
+            email_content["To"] = get_content_result.to_email
+            email_content["Subject"] = get_content_result.subject
```
yeah. tested it manually. 
I was passing this:
```
smtp_settings = SMTPServiceConfig(
    host='smtp.gmail.com',
    email_from=SMTPServiceConfigFrom(
        display_name,
        email
    ),
    auth=SMTPServiceConfigAuth(
        user=email,
        password=password,
    ),
    port=587,
    secure=True,
)
```

KShivendu replied:
how do you think we should do it?

KShivendu commented on supertokens_python/ingredients/emaildelivery/service/smtp.py:
```diff
@@ -0,0 +1,103 @@
+    @abstractmethod
+    async def send_raw_email(self,
+                             get_content_result: GetContentResult,
+                             config_from: SMTPServiceConfigFrom,
```
yes. It wasn't required. Thanks!

KShivendu commented on supertokens_python/recipe/emailpassword/emaildelivery/service/smtp/__init__.py:
```diff
@@ -0,0 +1,58 @@
+            return await self.email_verification_smtp_service.send_email(email_input, user_context)
+
+        content = await self.service_implementation.get_content(email_input, user_context)
+        send_raw_email_from = SMTPServiceConfigFrom(
+            self.config.smtp_settings.email_from.name,
+            self.config.smtp_settings.email_from.email
+        )
+        await self.service_implementation.send_raw_email(content, send_raw_email_from, user_context)
```
Done. Thanks!

KShivendu commented:
@sattvikc Other than your suggestions I did the following:
- Introduce is_html in GetContentResult and Transporter.send_email
- Include user_context in TypeEmailVerificationEmailDeliveryInput instead of passing as separate param
- Make EmailVerificationIngredients mandatory for recipes

All of these changes are based on the Node PR.

rishabhpoddar commented on supertokens_python/recipe/emailverification/emaildelivery/service/backward_compatibility/__init__.py:
```diff
@@ -0,0 +1,52 @@
+def default_create_and_send_custom_email(app_info: AppInfo) -> Callable[[User, str, Dict[str, Any]], Awaitable[None]]:
+    async def func(user: User, email_verification_url: str, _: Dict[str, Any]):
+        if ('SUPERTOKENS_ENV' in environ) and (environ['SUPERTOKENS_ENV'] == 'testing'):
+            return
```
need to add test that it comes in here

KShivendu replied:
Added tests in `tests/emailpassword/test_emaildelivery.py`

rishabhpoddar commented on supertokens_python/recipe/emailverification/emaildelivery/service/smtp/email_verify.py:
```diff
@@ -0,0 +1,37 @@
+def get_email_verify_email_content(email_input: TypeEmailVerificationEmailDeliveryInput) -> GetContentResult:
+    supertokens = Supertokens.get_instance()
+    app_name = supertokens.app_info.app_name
+    body = get_email_verify_email_html(app_name, email_input.user.email, email_input.email_verify_link)
```
need to add is_html as true

KShivendu replied:
done

rishabhpoddar commented on supertokens_python/recipe/emailverification/emaildelivery/service/smtp/implementation.py:
```diff
@@ -0,0 +1,30 @@
+# You may not use this file except in compliance with the License. You may
+# obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
```
rename this file to service_implementation.py

KShivendu replied:
Done.

rishabhpoddar commented on supertokens_python/recipe/emailverification/recipe.py:
```diff
@@ -45,19 +49,27 @@
     def __init__(self, recipe_id: str, app_info: AppInfo,
-                 config: ParentRecipeEmailVerificationConfig):
+                 config: ParentRecipeEmailVerificationConfig,
+                 ingredients: EmailVerificationIngredients,
```
Ingredients can be None too? If not, then you also need to make the emailpassword ingredients as non nullable

KShivendu replied:
Made both non-nullable.

rishabhpoddar commented on supertokens_python/recipe/emailverification/utils.py:
```diff
@@ -57,42 +52,52 @@ class ParentRecipeEmailVerificationConfig:
+        self.email_delivery = email_delivery
+        self.create_and_send_custom_email = create_and_send_custom_email
+
+        if create_and_send_custom_email:
+            warn("create_and_send_custom_email is depricated. Please use email delivery config instead")
```
We need to mark them as deprecated so that the IDE of the dev complains about it. Not this way

rishabhpoddar commented on supertokens_python/recipe/emailpassword/utils.py:
```diff
@@ -302,14 +296,34 @@ def validate_and_normalise_user_input(recipe: EmailPasswordRecipe, app_info: App
+            reset_password_using_token_feature=reset_password_using_token_feature,
+            email_verification_feature=email_verification_feature,
+        )
+        return EmailDeliveryConfigWithService(email_service, override=None)
```
You should not give override as None! Give it as what the user may have passed in `email_delivery_config`.

You need to add tests for this

KShivendu replied:
Fixed. Will add tests.

rishabhpoddar commented on supertokens_python/recipe/thirdpartyemailpassword/recipe.py:
```diff
@@ -96,7 +98,7 @@ def __init__(self, recipe_id: str, app_info: AppInfo,
         else:
             self.email_verification_recipe = EmailVerificationRecipe(recipe_id, app_info,
-                                                                     self.config.email_verification_feature)
+                                                                     self.config.email_verification_feature, EmailVerificationIngredients(None))
```
TODO: pass in email delivery ingredient in emailpassword instance as well.

KShivendu replied:
Done in the TPEP PR.
Tests are also passing in that PR so we can be sure that it's done.

KShivendu commented on supertokens_python/ingredients/emaildelivery/__init__.py:
```diff
@@ -0,0 +1,36 @@
+
+    def __init__(self, config: EmailDeliveryConfigWithService[_T]) -> None:
+        oi = DefaultEmailDeliveryIngredientImp[_T](config)
+        self.ingredient_interface_impl = oi if config.override is None else config.override(oi)
```
Done.

KShivendu commented on supertokens_python/ingredients/emaildelivery/__init__.py:
```diff
@@ -0,0 +1,36 @@
+_T = TypeVar('_T')
+
+
+class DefaultEmailDeliveryIngredientImp(EmailDeliveryInterface[_T]):
+    def __init__(self, config: EmailDeliveryConfigWithService[_T]) -> None:
+        self.config = config
+
+    async def send_email(self, email_input: _T, user_context: Dict[str, Any]) -> Any:
+        await self.config.service.send_email(email_input, user_context)
```
Done.

KShivendu commented on supertokens_python/ingredients/emaildelivery/__init__.py:
```diff
@@ -0,0 +1,36 @@
+    ingredient_interface_impl: EmailDeliveryInterface[_T]
+
+    def __init__(self, config: EmailDeliveryConfigWithService[_T]) -> None:
+        oi = DefaultEmailDeliveryIngredientImp[_T](config)
```
Done.

KShivendu commented on supertokens_python/ingredients/emaildelivery/service/smtp.py:
```diff
@@ -0,0 +1,133 @@
+        self.email = email
+
+
+class SMTPServiceConfig:
+    def __init__(
+        self, host: str, email_from: SMTPServiceConfigFrom,
+        port: int, secure: Union[bool, None] = None,
+        auth: Union[SMTPServiceConfigAuth, None] = None,
+        encryption: Literal['NONE', 'SSL', 'TLS'] = 'NONE',
+    ) -> None:
+        self.host = host
+        self.email_from = email_from
+        self.port = port
+        self.secure = secure
+        self.auth = auth
+        self.encryption = encryption
```
The node PR has the following:
```ts
export interface SMTPServiceConfig {
    host: string;
    from: {
        name: string;
        email: string;
    };
    port: number;
    secure?: boolean;
    password: string;
}
```
I think I have to remove `auth ` field and only take in the password. Will do that. 

> Also, why do you need encryption field?

There are 3 different modes for encryption. @sattvikc has introduced it in the golang PR and suggested the same here. Also, Django has in-built support for sending emails. I checked Django's code, it had encryption as a arg.

rishabhpoddar replied:
encryption is not there in golang PR: https://github.com/supertokens/supertokens-golang/pull/94/files#diff-522cc2f31e58cbd56423dea69a8bf5fc7067fca146961153258decbcd814f12aR7

rishabhpoddar replied:
So are we going to keep encryption or not? If not, then how are you automatically figuring out the value of it?

KShivendu replied:
@sattvikc might have considered removing it. 
> If not, then how are you automatically figuring out the value of it?
If user provides `secure=True` we start a TLS connection to the SMTP server. If it fails, we raise error. If use provides `secure=False`, we don't upgrade the connection to TLS and remain unencrypted.

Gmail, most common SMTP server, works seamslessly with secure=True. I have manually tested it and have also shared a video with you.

KShivendu commented on supertokens_python/ingredients/emaildelivery/service/smtp.py:
```diff
@@ -0,0 +1,133 @@
+        try:
+            if get_content_result.is_html:
+                email_content = MIMEText(get_content_result.body, "html")
+                email_content["From"] = config_from.email
```
Fixed. Nice catch. Thanks!

KShivendu commented on supertokens_python/ingredients/emaildelivery/service/smtp.py:
```diff
@@ -0,0 +1,133 @@
+        self.smtp_settings = smtp_settings
+
+    def connect(self):
+        try:
+            if self.smtp_settings.secure and self.smtp_settings.encryption == "SSL":
+                mail = smtplib.SMTP_SSL(self.smtp_settings.host, self.smtp_settings.port)
+                context = ssl.create_default_context()
+                if mail.has_extn("starttls"):
+                    mail.starttls(context=context)
+            else:
+                mail = smtplib.SMTP(self.smtp_settings.host, self.smtp_settings.port)
+
+            # only attempt TLS over non-secure connections
+            if not self.smtp_settings.secure and self.smtp_settings.encryption == "TLS":
+                mail.starttls()
+
+            if self.smtp_settings.auth:
+                mail.login(self.smtp_settings.auth.user, self.smtp_settings.auth.password)
+
+            mail.ehlo_or_helo_if_needed()
+            return mail
```
I took Django's code as a baseline adding encryption parameter to this. And I manually tested with one of the encryption modes. (TLS)

KShivendu commented on supertokens_python/recipe/emailpassword/emaildelivery/service/backward_compatibility/__init__.py:
```diff
@@ -0,0 +1,120 @@
+
+                create_and_send_custom_email = create_and_send_custom_email_wrapper
+
+            self.email_verification_feature = email_verification_feature
```
I agree. Removed.

KShivendu commented on supertokens_python/recipe/emailpassword/emaildelivery/service/smtp/password_reset_email.html:
```diff
@@ -0,0 +1,940 @@
+<!doctype html>
+<html xmlns="http://www.w3.org/1999/xhtml" xmlns:v="urn:schemas-microsoft-com:vml"
+    xmlns:o="urn:schemas-microsoft-com:office:office">
+
```
Will update towards the end. When we are about to merge this one.

KShivendu replied:
Moved to a python file `password_reset_email.py`

KShivendu commented on supertokens_python/recipe/emailpassword/emaildelivery/service/smtp/password_reset_implementation.py:
```diff
@@ -0,0 +1,38 @@
+    supertokens = Supertokens.get_instance()
+    app_name = supertokens.app_info.app_name
+    body = get_password_reset_email_html(app_name, email_input.user.email, email_input.password_reset_link)
+    content_result = GetContentResult(body, "Password reset instructions", email_input.user.email)
```
Done. It was set to True by default. But as you have suggested, default is now `False`.

rishabhpoddar replied:
No. there should not be a default at all

KShivendu commented on supertokens_python/ingredients/emaildelivery/service/smtp.py:
```diff
@@ -0,0 +1,133 @@
+
+
+class GetContentResult:
+    def __init__(self, body: str, subject: str, to_email: str, is_html: bool = True) -> None:
```
Done.

KShivendu commented on supertokens_python/recipe/emailpassword/recipe.py:
```diff
@@ -49,26 +59,45 @@
+        self.recipe_implementation = recipe_implementation if self.config.override.functions is None else \
+            self.config.override.functions(recipe_implementation)
+
+        self.recipe_implementation = cast(RecipeImplementation, self.recipe_implementation)
```
`get_email_delivery_config` looks for `emailpassword.recipe_implementation.RecipeImplementation`

But the type checkers (like mypy) feel that `recipe_implementation` can be `RecipeInterface`, which for them is different from `RecipeImplementation`

After thinking a little bit, I've now changed it to:
```python
if not isinstance(self.recipe_implementation, RecipeImplementation):
            raise Exception("Should never come here")
```
instead of `self.recipe_implementation = cast(RecipeImplementation, self.recipe_implementation)`

KShivendu replied:
Nevermind. The type of `get_email_delivery_config` is now `Callable[[RecipeInterface], EmailDeliveryConfigWithService[TypeEmailPasswordEmailDeliveryInput]]`. We can/should use `RecipeInterface` not `RecipeImplementation`. 
No need of this:
```
if not isinstance(self.recipe_implementation, RecipeImplementation):
            raise Exception("Should never come here")
```

KShivendu commented on supertokens_python/recipe/emailpassword/recipe.py:
```diff
@@ -49,26 +59,45 @@
         if email_verification_recipe is not None:
             self.email_verification_recipe = email_verification_recipe
         else:
+            ev_email_delivery_ingredient = cast(
+                EmailDeliveryIngredient[TypeEmailVerificationEmailDeliveryInput],
+                self.email_delivery_ingredient
+            )
```
```
Type "EmailDeliveryIngredient[TypeEmailPasswordEmailDeliveryInput]" cannot be assigned to
type "EmailDeliveryIngredient[TypeEmailVerificationEmailDeliveryInput]
```

rishabhpoddar replied:
Using cast for email delivery ingredient is fine. Nothing else please

KShivendu commented on supertokens_python/recipe/emailpassword/utils.py:
```diff
@@ -185,11 +172,13 @@ def __init__(self,
         self.form_fields_for_generate_token_form = form_fields_for_generate_token_form
         self.get_reset_password_url = get_reset_password_url
         self.create_and_send_custom_email = create_and_send_custom_email
+        if create_and_send_custom_email:
+            warn("create_and_send_custom_email is deprecated. Please use email delivery config instead")
```
Couldn't find anything like that. Mostly it's libraries that give warning or fail if a particular function is used. Example: [Deprecated](https://pypi.org/project/Deprecated/), [Deprecation](https://deprecation.readthedocs.io/en/latest/)

Will looking for it again.

KShivendu commented on supertokens_python/recipe/emailpassword/utils.py:
```diff
@@ -213,8 +202,10 @@ def validate_and_normalise_reset_password_using_token_config(app_info: AppInfo,
+    if config.create_and_send_custom_email is not None:
+        create_and_send_custom_email = config.create_and_send_custom_email
+    else:
+        create_and_send_custom_email = default_create_and_send_custom_email(app_info)
```
Logic is the same. I did this change 2 months ago. I must have done it make it more readable (this line is quite long)

KShivendu replied:
Restored it.

KShivendu commented on supertokens_python/recipe/emailpassword/utils.py:
```diff
@@ -302,14 +296,34 @@ def validate_and_normalise_user_input(recipe: EmailPasswordRecipe, app_info: App
         sign_up_feature = InputSignUpFeature()
+
+    def get_email_delivery_config(
+        ep_recipe: RecipeImplementation,
```
Fixed.

KShivendu replied:
Also fixed it in `BackwardCompatibilityService`

KShivendu commented on supertokens_python/recipe/emailverification/utils.py:
```diff
@@ -57,42 +52,52 @@ class ParentRecipeEmailVerificationConfig:
+        if email_service is None:
+            email_service = BackwardCompatibilityService(app_info, config.create_and_send_custom_email)
+
+        return EmailDeliveryConfigWithService(email_service, override=None)
```
Fixed. Will add tests.

KShivendu commented on supertokens_python/ingredients/emaildelivery/service/smtp.py:
```diff
@@ -0,0 +1,133 @@
+            mail.ehlo_or_helo_if_needed()
+            return mail
+        except Exception as e:
+            log_debug_message("Couldn't connect to the SMTP server: %s", e)
```
Done.

KShivendu commented on supertokens_python/ingredients/emaildelivery/service/smtp.py:
```diff
@@ -0,0 +1,133 @@
+                         _: Dict[str, Any]) -> None:
+        connection = self.connect()
+        if connection is None:
+            return
```
Done.

KShivendu commented:
@rishabhpoddar other than your suggestions, I also did the following:
- Renamed variables like `email_from` to `from_`, `email_input` to `input_`. Both `input` and `from` are reserved keywords. This change made the code more similar to the node PR. 
- Renamed `service` dir to `services`. Again to make it consistent with the node PR.

KShivendu commented on supertokens_python/ingredients/emaildelivery/service/smtp.py:
```diff
@@ -0,0 +1,133 @@
+                email_content["Subject"] = get_content_result.subject
+                connection.sendmail(config_from.email, get_content_result.to_email, email_content.as_string())
+            else:
+                connection.sendmail(config_from.email, get_content_result.to_email, get_content_result.body)
```
Tried fix it based on what I found on the internet. 
```
email_content["From"] = f"{sender.name} <{sender.email}>"
```

Will take note of this while manually testing.

KShivendu commented on supertokens_python/recipe/emailverification/emaildelivery/__init__.py:
```diff
@@ -0,0 +1,13 @@
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+# License for the specific language governing permissions and limitations
+# under the License.
```
Done.

KShivendu commented on supertokens_python/recipe/emailpassword/emaildelivery/services/smtp/password_reset_email.html:
```diff
@@ -0,0 +1,940 @@
+<!doctype html>
```
Done. Moved to a seperate .py file so that we know when only the template has changed. It keeps the code as well as diff clean.

KShivendu commented on supertokens_python/ingredients/emaildelivery/services/smtp.py:
```diff
@@ -0,0 +1,136 @@
+        self, host: str, from_: SMTPServiceConfigFrom,
+        port: int, secure: Union[bool, None] = None,
+        auth: Union[SMTPServiceConfigAuth, None] = None,
+        encryption: Literal['NONE', 'SSL', 'TLS'] = 'NONE',
```
@rishabhpoddar 
Removed the encryption parameter and avoiding taking in the from_email again in `auth` field. (Just like the Node PR)

I also shared a video with you where I tested transporter's logic with a local SMTP server as well as a Gmail account.

rishabhpoddar commented on supertokens_python/ingredients/emaildelivery/services/smtp.py:
```diff
@@ -0,0 +1,121 @@
+    async def send_email(self, input_: GetContentResult,
+                         _: Dict[str, Any]) -> None:
+        connection = self._connect()
+        if connection is None:
+            raise Exception("Couldn't connect to the SMTP server.")
```
You don't need this check anymore right?

KShivendu replied:
yeah. not required anymore. removed.

KShivendu commented:
@rishabhpoddar Along with the changes suggested by you. I also did the following:
- Fixed other syncio functions that had `async` in front them and/or were initializing user_context before passing to the `async` implementation.

KShivendu commented on supertokens_python/ingredients/emaildelivery/services/smtp.py:
```diff
@@ -0,0 +1,121 @@
+                email_content["From"] = from_addr
+                email_content["To"] = input_.to_email
+                email_content["Subject"] = input_.subject
+                connection.sendmail(from_.email, input_.to_email, email_content.as_string())
+            else:
+                connection.sendmail(from_addr, input_.to_email, input_.body)
```
Yes. This is what the in-built `smtplib` provides. It isn't a async function.

rishabhpoddar replied:
Are you sure there is no async version of this? Cause if this is blocking, then it will block the webserver from serving any requests while this is running..

KShivendu commented on supertokens_python/recipe/emailpassword/emaildelivery/services/smtp/service_implementation/__init__.py:
```diff
@@ -0,0 +1,50 @@
+    async def get_content(self, input_: TypeEmailPasswordEmailDeliveryInput) -> GetContentResult:
+        if isinstance(input_, TypeEmailVerificationEmailDeliveryInput):
+            return await self.ev_get_content(input_)
+        return get_password_reset_email_content(input_)
```
await

KShivendu commented on supertokens_python/ingredients/emaildelivery/interfaces.py:
```diff
@@ -0,0 +1,27 @@
+
```
Done

KShivendu commented on supertokens_python/ingredients/emaildelivery/interfaces.py:
```diff
@@ -0,0 +1,27 @@
+
+from typing import Any, Dict, Generic, TypeVar, Union
```
Done.

KShivendu commented on supertokens_python/ingredients/emaildelivery/interfaces.py:
```diff
@@ -0,0 +1,27 @@
+        self.config = config
+
+    def override(self, oi: EmailDeliveryInterface[_T]) -> EmailDeliveryInterface[_T]:
+        return oi
```
Done.

KShivendu commented on supertokens_python/ingredients/emaildelivery/service/smtp.py:
```diff
@@ -0,0 +1,112 @@
+    default_impl = getDefaultEmailServiceImplementation(transporter, input_send_raw_email)
+    # impl = config.override if config.override is not None else default_impl
+
+    return default_impl
```
Done.

KShivendu commented on supertokens_python/recipe/emailverification/emaildelivery/services/backward/__init__.py:
```diff
@@ -0,0 +1,37 @@
+
```
Calling it `backward_compatibility`

KShivendu commented on supertokens_python/recipe/emailverification/interfaces.py:
```diff
@@ -224,6 +227,17 @@ def __init__(self):
 
+class TypeEmailVerificationEmailDeliveryInput:
+    def __init__(
+        self, user: User,
```
Fixed.

KShivendu commented on supertokens_python/recipe/emailverification/interfaces.py:
```diff
@@ -224,6 +227,17 @@ def __init__(self):
+        self, user: User,
+        email_verify_link: str, user_context: Dict[str, Any]
+    ) -> None:
+        self.type: str = "EMAIL_VERIFICATION"
```
Removed.

KShivendu commented on supertokens_python/recipe/emailverification/emaildelivery/services/backward/__init__.py:
```diff
@@ -0,0 +1,37 @@
+
+    async def send_email(self, email_input: Union[TypeEmailVerificationEmailDeliveryInput, None]) -> Any:
+        if email_input is None:
+            return
```
Removed.

KShivendu commented on supertokens_python/recipe/emailpassword/interfaces.py:
```diff
@@ -197,13 +200,16 @@ async def update_email_or_password(self, user_id: str, email: Union[str, None],
-                 email_verification_recipe_implementation: EmailVerificationRecipeInterface):
+                 email_verification_recipe_implementation: EmailVerificationRecipeInterface,
+                 email_delivery: Union[EmailDeliveryIngredient[TypeEmailPasswordEmailDeliveryInput], None] = None,
+                 ):
```
Fixed.

KShivendu commented on supertokens_python/ingredients/emaildelivery/types.py:
```diff
@@ -0,0 +1,37 @@
+        override: Union[Callable[[EmailDeliveryInterface[_T]], EmailDeliveryInterface[_T]], None] = None,
+    ) -> None:
+        self.service = service
+        self.override = override
```
Done. Introduced `EmailDeliveryConfigWithService`

KShivendu commented on supertokens_python/recipe/emailpassword/emaildelivery/services/backward_compatibility/__init__.py:
```diff
@@ -0,0 +1,116 @@
+                 reset_password_using_token_feature: Union[InputResetPasswordUsingTokenFeature, None] = None,
+                 email_verification_feature: Union[InputEmailVerificationConfig, None] = None,
+                 ) -> None:
+        self.app_info = app_info
```
yeah `app_info` wasn't required. but we need `recipe_interface_impl`

KShivendu commented on supertokens_python/ingredients/emaildelivery/services/smtp.py:
```diff
@@ -0,0 +1,121 @@
+
+
+class GetContentResult:
+    def __init__(self, body: str, subject: str, to_email: str, is_html: bool = False) -> None:
```
Fixed.

KShivendu commented on supertokens_python/recipe/thirdpartypasswordless/recipe.py:
```diff
@@ -138,7 +140,7 @@ async def is_email_verified(user_id: str, email: str, user_context: Dict[str, An
                 self.config.email_verification_feature.override.functions = email_verification_override
             self.email_verification_recipe = EmailVerificationRecipe(recipe_id, app_info,
-                                                                     self.config.email_verification_feature)
+                                                                     self.config.email_verification_feature, EmailVerificationIngredients(None))
```
Done in the Tpples PR.

KShivendu replied:
Tests are also passing in that PR so we can be sure that it's done.

KShivendu replied:
See https://github.com/supertokens/supertokens-python/pull/162/files#diff-2a3e02d971230ab8f514c295f7e512b9e85d8366d51df80557bac3f260f7d89aR171

KShivendu commented on supertokens_python/recipe/thirdparty/recipe.py:
```diff
@@ -58,7 +60,7 @@ def __init__(self, recipe_id: str, app_info: AppInfo,
         else:
             self.email_verification_recipe = EmailVerificationRecipe(recipe_id, app_info,
-                                                                     self.config.email_verification_feature)
+                                                                     self.config.email_verification_feature, EmailVerificationIngredients(None))
```
Done. See https://github.com/supertokens/supertokens-python/pull/163/files#diff-4c886eb7113745dd43a1793cd02dfe0a43000b14434170388026bf6800485b9cR88

rishabhpoddar commented on supertokens_python/recipe/emailpassword/syncio/__init__.py:
```diff
@@ -31,7 +31,7 @@ def verify_email_using_token(token: str, user_context: Union[None, Dict[str, Any
     return sync(verify_email_using_token(token, user_context))
 
 
-async def unverify_email(user_id: str, user_context: Union[None, Dict[str, Any]] = None):
```
This fix should go in another PR to the main x.y branch as it is unrelated to this PR.

KShivendu replied:
Done.

rishabhpoddar replied:
You need to make another PR fixing this (to the latest X.Y branch).

KShivendu replied:
Done.

rishabhpoddar commented on supertokens_python/recipe/emailpassword/syncio/__init__.py:
```diff
@@ -42,8 +42,8 @@ def is_email_verified(user_id: str, user_context: Union[None, Dict[str, Any]] =
     return sync(is_email_verified(user_id, user_context))
 
 
-async def update_email_or_password(user_id: str, email: Union[str, None] = None,
```
This fix should go in another PR to the main x.y branch as it is unrelated to this PR.

rishabhpoddar replied:
You need to make another PR fixing this (to the latest X.Y branch).

KShivendu replied:
Created new PR https://github.com/supertokens/supertokens-python/pull/168/files

rishabhpoddar commented on supertokens_python/recipe/emailpassword/syncio/__init__.py:
```diff
@@ -79,12 +79,17 @@ def sign_in(email: str, password: str, user_context: Union[None, Dict[str, Any]]
     return sync(sign_in(email, password, user_context))
 
 
-async def sign_up(email: str, password: str, user_context: Union[None, Dict[str, Any]] = None):
```
This fix should go in another PR to the main x.y branch as it is unrelated to this PR.

KShivendu replied:
Fixed in https://github.com/supertokens/supertokens-python/pull/168/files

rishabhpoddar commented on supertokens_python/recipe/emailpassword/syncio/__init__.py:
```diff
@@ -79,12 +79,17 @@ def sign_in(email: str, password: str, user_context: Union[None, Dict[str, Any]]
     return sync(sign_up(email, password, user_context))
 
 
-async def revoke_email_verification_token(user_id: str, user_context: Union[None, Dict[str, Any]] = None):
```
This fix should go in another PR to the main x.y branch as it is unrelated to this PR.

rishabhpoddar replied:
You need to make another PR fixing this (to the latest X.Y branch).

KShivendu replied:
Fixed in https://github.com/supertokens/supertokens-python/pull/168/files

rishabhpoddar commented on supertokens_python/recipe/emailpassword/emaildelivery/services/backward_compatibility/__init__.py:
```diff
@@ -41,16 +43,20 @@ def default_create_and_send_custom_email(
                 'appName': app_info.app_name,
                 'passwordResetURL': password_reset_url_with_token
             }
             async with AsyncClient() as client:
```
```suggestion
        data = {
                'email': user.email,
                'appName': app_info.app_name,
                'passwordResetURL': password_reset_url_with_token
            }
        try:
            async with AsyncClient() as client:
```

KShivendu replied:
We are passing data to `handle_httpx_client_exceptions` outside try/except block. So it needs to be defined beforehand.

rishabhpoddar replied:
See my comment again.. i have moved it above the try block

KShivendu replied:
Fixed.

rishabhpoddar commented on supertokens_python/recipe/emailpassword/emaildelivery/services/backward_compatibility/__init__.py:
```diff
@@ -41,16 +43,20 @@ def default_create_and_send_custom_email(
-        except Exception:
-            pass
+                resp = await client.post('https://api.supertokens.io/0/st/auth/password/reset', json=data, headers={'api-version': '0'})  # type: ignore
+                resp.raise_for_status()
```
What does `resp.raise_for_status()` mean?

KShivendu replied:
If we get resp.status_code <= 599 or resp.status_code >= 400, it will raise an error (Like Axios in node)

rishabhpoddar commented on supertokens_python/recipe/emailverification/emaildelivery/services/backward_compatibility/__init__.py:
```diff
@@ -18,21 +18,29 @@
             return
+        data = {}
         try:
+            data = {'email': user.email, 'appName': app_info.app_name, 'emailVerifyURL': email_verification_url}
```
```suggestion
        data = {'email': user.email, 'appName': app_info.app_name, 'emailVerifyURL': email_verification_url}
        try:
```

KShivendu replied:
We are passing `data` to `handle_httpx_client_exceptions` outside try/except block. So it needs to be defined beforehand.

rishabhpoddar replied:
See my comment again. It's above the try block

rishabhpoddar commented on supertokens_python/recipe/emailverification/emaildelivery/services/backward_compatibility/__init__.py:
```diff
@@ -18,21 +18,29 @@
-        except Exception:
-            pass
+                resp = await client.post('https://api.supertokens.io/0/st/auth/email/verify', json=data, headers={'api-version': '0'})  # type: ignore
+                resp.raise_for_status()
```
What is `resp.raise_for_status()` mean?

KShivendu replied:
If we get resp.status_code <= 599 or resp.status_code >= 400, it will raise an error (Like Axios in node)

rishabhpoddar commented on supertokens_python/utils.py:
```diff
@@ -179,3 +182,14 @@ def frontend_has_interceptor(request: BaseRequest) -> bool:
+
+
+def handle_httpx_client_exceptions(e: Exception, input_: Dict[str, Any]):
+    if isinstance(e, HTTPStatusError):
+        res: Response = e.response  # type: ignore
+        log_debug_message("Error status: %s", res.status_code)  # type: ignore
+        log_debug_message("Error response: %s", res.json())
+    else:
+        log_debug_message("Error: %s", e)
+        log_debug_message("Logging the input:")
+        log_debug_message("%s", json.dumps(input_))
```
why are you logging the input in only one case?

KShivendu replied:
Fixed.

rishabhpoddar commented on supertokens_python/recipe/thirdpartyemailpassword/recipe.py:
```diff
@@ -109,6 +113,7 @@ def apis_override_email_password(_: EmailPasswordAPIInterface) -> EmailPasswordA
             self.email_password_recipe = email_password_recipe
         else:
             self.email_password_recipe = EmailPasswordRecipe(recipe_id, app_info,
+                                                             EmailPasswordIngredients(None),
```
You should not pass none to this

KShivendu replied:
It's fine. It's passed in the PR for tpep https://github.com/supertokens/supertokens-python/pull/163/files#diff-1d423b5d85fc2ebb83925be074e6758f869042c73d9f9a8ac33b96d38757b71eR146

rishabhpoddar commented on supertokens_python/recipe/thirdpartyemailpassword/recipe.py:
```diff
@@ -96,7 +100,7 @@ def __init__(self, recipe_id: str, app_info: AppInfo,
         else:
             self.email_verification_recipe = EmailVerificationRecipe(recipe_id, app_info,
-                                                                     self.config.email_verification_feature)
+                                                                     self.config.email_verification_feature, EmailVerificationIngredients(None))
```
Do not pass none to this

KShivendu replied:
It's fine for now. It's passed in the relevant PR for the tpep recipe https://github.com/supertokens/supertokens-python/pull/163/files#diff-1d423b5d85fc2ebb83925be074e6758f869042c73d9f9a8ac33b96d38757b71eR132

rishabhpoddar commented on supertokens_python/utils.py:
```diff
@@ -175,3 +179,43 @@ def execute_async(mode: str, func: Callable[[], Coroutine[Any, Any, None]]):
+    log_debug_message("%s", json.dumps(input_))
+
+
+def humanize_time(ms: int) -> str:
```
Have you written tests for this? Where are they?

KShivendu replied:
Removed it. It's more sensible to keep it in https://github.com/supertokens/supertokens-python/pull/162/

rishabhpoddar commented on supertokens_python/recipe/emailpassword/emaildelivery/services/backward_compatibility/__init__.py:
```diff
@@ -0,0 +1,120 @@
+        if reset_password_using_token_feature and reset_password_using_token_feature.create_and_send_custom_email is not None:
+            reset_password_feature_send_email_func = reset_password_using_token_feature.create_and_send_custom_email
+
+        self.reset_password_feature_send_email_func = reset_password_feature_send_email_func
```
No. Doesn't make any difference.

rishabhpoddar commented on supertokens_python/recipe/emailpassword/emaildelivery/services/smtp/__init__.py:
```diff
@@ -0,0 +1,53 @@
+
+
+class SMTPService(EmailDeliveryInterface[TypeEmailPasswordEmailDeliveryInput]):
+    service_implementation: ServiceInterface[TypeEmailPasswordEmailDeliveryInput]
```
No need

rishabhpoddar commented on supertokens_python/recipe/emailpassword/emaildelivery/services/smtp/__init__.py:
```diff
@@ -0,0 +1,53 @@
+            smtp_settings=config.smtp_settings,
+            override=lambda _: EmailVerificationServiceImpl(self.service_implementation)
+        )
+        self.email_verification_smtp_service = EmailVerificationSMTPService(ev_config)
```
No need

rishabhpoddar commented on supertokens_python/recipe/emailpassword/emaildelivery/services/smtp/service_implementation/email_verification_implementation.py:
```diff
@@ -0,0 +1,33 @@
+# Copyright (c) 2021, VRAI Labs and/or its affiliates. All rights reserved.
```
no need

rishabhpoddar commented on supertokens_python/recipe/emailpassword/emaildelivery/services/backward_compatibility/__init__.py:
```diff
@@ -0,0 +1,120 @@
+        InputEmailVerificationConfig, InputResetPasswordUsingTokenFeature)
+
+
+def default_create_and_send_custom_email(
```
no need

rishabhpoddar commented on supertokens_python/utils.py:
```diff
@@ -175,3 +179,44 @@ def execute_async(mode: str, func: Callable[[], Coroutine[Any, Any, None]]):
+
+    if input_ is not None:
+        log_debug_message("Logging the input:")
+        log_debug_message("%s", json.dumps(input_))
```
both are weird. the \n makes it very unreadable.

rishabhpoddar replied:
but it's just a debug log.. so no need to change anything

rishabhpoddar commented on supertokens_python/recipe/emailpassword/emaildelivery/services/__init__.py:
```diff
@@ -0,0 +1,18 @@
+# License for the specific language governing permissions and limitations
+# under the License.
+
+from .smtp import EmailDeliverySMTPConfig, SMTPService
```
Huh?

rishabhpoddar commented on supertokens_python/recipe/emailpassword/utils.py:
```diff
@@ -185,12 +169,15 @@ def __init__(self,
                  form_fields_for_generate_token_form: List[NormalisedFormField],
                  get_reset_password_url: Callable[[User, Dict[str, Any]], Awaitable[str]],
-                 create_and_send_custom_email: Callable[[User, str, Dict[str, Any]], Awaitable[None]]):
+                 create_and_send_custom_email: Union[Callable[[User, str, Dict[str, Any]], Awaitable[None]], None] = None):
```
You need to remove this from the normalised type. Same goes for email verification normalised type.

rishabhpoddar replied:
@KShivendu missed this?

rishabhpoddar commented on supertokens_python/recipe/passwordless/smsdelivery/services/twilio/pless_sms.py:
```diff
@@ -0,0 +1,76 @@
+# Copyright (c) 2021, VRAI Labs and/or its affiliates. All rights reserved.
```
ok

KShivendu replied:
done.

rishabhpoddar commented on supertokens_python/recipe/thirdpartyemailpassword/emaildelivery/services/backward_compatibility/__init__.py:
```diff
@@ -0,0 +1,86 @@
+                 reset_password_using_token_feature: Union[InputResetPasswordUsingTokenFeature, None] = None,
+                 email_verification_feature: Union[InputEmailVerificationConfig, None] = None,
+                 ) -> None:
+        self.recipe_interface_impl = recipe_interface_impl
```
ok

KShivendu replied:
done.

rishabhpoddar commented on supertokens_python/recipe/thirdpartyemailpassword/emaildelivery/services/backward_compatibility/__init__.py:
```diff
@@ -0,0 +1,86 @@
+
+    async def send_email(self, input_: TypeThirdPartyEmailPasswordEmailDeliveryInput, user_context: Dict[str, Any]) -> None:
+        if isinstance(input_, TypeEmailVerificationEmailDeliveryInput):
+            return await self.ev_backward_compatiblity_service.send_email(input_, user_context)
```
ok do this

KShivendu replied:
done.

rishabhpoddar commented on supertokens_python/recipe/thirdpartyemailpassword/recipe.py:
```diff
@@ -92,47 +112,67 @@ def __init__(self, recipe_id: str, app_info: AppInfo,
             self.email_password_recipe = email_password_recipe
         else:
+            def func_override_email_password(_: EmailPasswordRecipeInterface):
+                return EmailPasswordRecipeImplementation(self.recipe_implementation)
```
do this.

KShivendu replied:
done.

rishabhpoddar commented on supertokens_python/recipe/thirdpartyemailpassword/utils.py:
```diff
@@ -128,7 +138,8 @@ def validate_and_normalise_user_input(
         override: Union[InputOverrideConfig, None] = None,
-        providers: Union[List[Provider], None] = None
+        providers: Union[List[Provider], None] = None,
+        email_delivery_config: Union[EmailDeliveryConfig[TypeThirdPartyEmailPasswordEmailDeliveryInput], None] = None,
```
no need.

KShivendu replied:
done.

rishabhpoddar commented on supertokens_python/recipe/thirdpartypasswordless/emaildelivery/services/backward_compatibility/__init__.py:
```diff
@@ -0,0 +1,81 @@
+    def __init__(self,
+                 app_info: AppInfo,
+                 recipe_interface_impl: RecipeInterface,
+                 create_and_send_custom_email: Union[Callable[
```
no need to do this

rishabhpoddar commented on supertokens_python/recipe/thirdpartypasswordless/emaildelivery/services/smtp/__init__.py:
```diff
@@ -0,0 +1,64 @@
+    service_implementation: ServiceInterface[TypeThirdPartyPasswordlessEmailDeliveryInput]
+
+    def __init__(self, config: EmailDeliverySMTPConfig[TypeThirdPartyPasswordlessEmailDeliveryInput]) -> None:
+        self.transporter = Transporter(config.smtp_settings)
```
no need to do this

rishabhpoddar commented on supertokens_python/recipe/thirdpartypasswordless/emaildelivery/services/smtp/__init__.py:
```diff
@@ -0,0 +1,64 @@
+        self.transporter = Transporter(config.smtp_settings)
+
+        oi = ServiceImplementation(self.transporter)
+        self.service_implementation = oi if config.override is None else config.override(oi)
```
ok do this

KShivendu replied:
done.

rishabhpoddar commented on supertokens_python/recipe/thirdpartypasswordless/smsdelivery/services/twilio/__init__.py:
```diff
@@ -0,0 +1,38 @@
+    def __init__(self, config: SMSDeliveryTwilioConfig[TypeThirdPartyPasswordlessSmsDeliveryInput]) -> None:
+        self.pless_twilio_service = PlessTwilioService(config)
+
+    async def send_sms(self, input_: TypePasswordlessSmsDeliveryInput, user_context: Dict[str, Any]) -> None:
```
ok do this.

KShivendu replied:
done.

rishabhpoddar commented on CHANGELOG.md:
```diff
@@ -6,6 +6,119 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
+-   `SMTP` service integration for `email_delivery` ingredient.
+-   `Supertokens` service integration for `sms_delivery` ingredient.
+
+### Breaking Change
+
+-   `user` object passed for password-reset and email-verification send email functions will only have `id` and `email` properties. `time_joined` property has been removed.
```
Is this accurate?
